server/src/main/java/org/elasticsearch/search/profile/query/ProfileCollectorManager.java

```diff
@@ -8,54 +8,112 @@
  * in an {@link InternalProfileCollector}. It delegates all the profiling to the generated collectors via {@link #getCollectorTree()}
- * and joins them up when its {@link #reduce} method is called. The profile result can
+ * and joins the different collector trees together when its {@link #reduce} method is called.
+ * Supports optionally providing sub-collector managers for top docs as well as aggs collection, so that each
+ * {@link InternalProfileCollector} created is provided with the corresponding sub-collectors that are children of the top-level collector.
+ * @param <T> the return type of the wrapped collector manager, which the reduce method returns.
  */
 public final class ProfileCollectorManager<T> implements CollectorManager<InternalProfileCollector, T> {
 
-    private final CollectorManager<Collector, T> collectorManager;
+    private final CollectorManager<? extends Collector, T> collectorManager;
     private final String reason;
+    private final ProfileCollectorManager<?> topDocsSubCollectorManager;
+    private final ProfileCollectorManager<?> aggsSubCollectorManager;
+    // this is a bit of a hack: it allows us to retrieve the last collector that newCollector has returned for sub-collector managers,
+    // so that we can provide them to InternalProfileCollector's constructor as children.
+    private InternalProfileCollector profileCollector;
+
     private CollectorResult collectorTree;
 
-    @SuppressWarnings("unchecked")
     public ProfileCollectorManager(CollectorManager<? extends Collector, T> collectorManager, String reason) {
-        this.collectorManager = (CollectorManager<Collector, T>) collectorManager;
+        this(collectorManager, reason, null, null);
+    }
+
+    public ProfileCollectorManager(
+        CollectorManager<? extends Collector, T> collectorManager,
+        String reason,
+        ProfileCollectorManager<?> topDocsSubCollectorManager,
+        ProfileCollectorManager<?> aggsSubCollectorManager
+    ) {
+        this.collectorManager = collectorManager;
         this.reason = reason;
+        this.topDocsSubCollectorManager = topDocsSubCollectorManager;
+        this.aggsSubCollectorManager = aggsSubCollectorManager;
     }
 
     @Override
     public InternalProfileCollector newCollector() throws IOException {
-        return new InternalProfileCollector(collectorManager.newCollector(), reason);
+        Collector collector = collectorManager.newCollector();
+        if (aggsSubCollectorManager == null && topDocsSubCollectorManager == null) {
+            profileCollector = new InternalProfileCollector(collector, reason);
+        } else if (aggsSubCollectorManager == null) {
+            assert topDocsSubCollectorManager.profileCollector != null;
+            profileCollector = new InternalProfileCollector(collector, reason, topDocsSubCollectorManager.profileCollector);
+        } else {
```

Comment:
There seems to be an assumption here that topDocsSubCollectorManager can never be "null" alone. Should we make that part of the docs and check in the constructor that this is the case?
i.e 
* both sub-collectormanagers "null" -> okay
* only aggsSubCollectorManager == null -> okay
* both not "null" -> okay
* throw in the last case?

Reply:
yea that's a good idea.

Reply:
I added an assertion.